lib/fsi.go

```diff
@@ -197,6 +197,11 @@ func (m *FSIMethods) Checkout(p *CheckoutParams, out *string) (err error) {
 	}
 	log.Debugf("Checkout loaded dataset %q", ref)
 
+	// Fail early if link already exists
+	if _, err := m.inst.fsi.HasLink(p.Ref); err != nil {
+		return err
+	}
```

Comment:
Bailing out early is good for now. But just putting this thought here for documentation: what would be even better in the future would be to use the `rollback` functionality instead, to undo previous filesystem modifications once an error gets hit. The `init` command uses it, but `checkout` doesn't. After all, it's possible that a later step fails, like `fsi.WriteComponents`, in which case we don't undo the directory creation. Using a rollback lets us be sure that an error anywhere restores a sane state to the user's machine.